crates/astria-sequencer/src/assets/state_ext.rs

```diff
@@ -0,0 +1,627 @@
+        // let mut fees: Vec<(asset::Id, u128)> = Vec::new();
+        let mut fees = Vec::new();
+
+        let mut stream =
+            std::pin::pin!(self.nonverifiable_prefix_raw(BLOCK_FEES_PREFIX.as_bytes()));
+        while let Some(Ok((key, value))) = stream.next().await {
+            // if the key isn't of the form `block_fees/{asset_id}`, then we have a bug
+            // in `put_block_fees`
+            let suffix = key
+                .strip_prefix(BLOCK_FEES_PREFIX.as_bytes())
+                .expect("prefix must always be present");
+            let asset = std::str::from_utf8(suffix)
+                .context("key suffix was not utf8 encoded; this should not happen")?
+                .parse::<crate::storage_keys::hunks::Asset>()
+                .context("failed to parse storage key suffix as address hunk")?
+                .get();
+
+            let Ok(bytes): Result<[u8; 16], _> = value.try_into() else {
+                bail!("failed turning raw block fees bytes into u128; not 16 bytes?");
+            };
+
+            fees.push((asset, u128::from_be_bytes(bytes)));
+        }
+
+        Ok(fees)
+    }
+
+    #[instrument(skip_all)]
+    async fn is_allowed_fee_asset<TAsset>(&self, asset: TAsset) -> Result<bool>
+    where
+        TAsset: Into<asset::IbcPrefixed> + std::fmt::Display + Send,
+    {
+        Ok(self
+            .nonverifiable_get_raw(fee_asset_key(asset).as_bytes())
+            .await
+            .context("failed to read raw fee asset from state")?
+            .is_some())
+    }
+
+    #[instrument(skip_all)]
+    async fn get_allowed_fee_assets(&self) -> Result<Vec<asset::IbcPrefixed>> {
+        let mut assets = Vec::new();
+
+        let mut stream = std::pin::pin!(self.nonverifiable_prefix_raw(FEE_ASSET_PREFIX.as_bytes()));
+        while let Some(Ok((key, _))) = stream.next().await {
+            // if the key isn't of the form `fee_asset/{asset_id}`, then we have a bug
+            // in `put_allowed_fee_asset`
+            let suffix = key
+                .strip_prefix(FEE_ASSET_PREFIX.as_bytes())
+                .expect("prefix must always be present");
+            let asset = std::str::from_utf8(suffix)
+                .context("key suffix was not utf8 encoded; this should not happen")?
+                .parse::<crate::storage_keys::hunks::Asset>()
+                .context("failed to parse storage key suffix as address hunk")?
+                .get();
+            assets.push(asset);
+        }
+
+        Ok(assets)
+    }
+}
+
+impl<T: ?Sized + StateRead> StateReadExt for T {}
+
+#[async_trait]
+pub(crate) trait StateWriteExt: StateWrite {
+    #[instrument(skip_all)]
+    fn put_native_asset(&mut self, asset: &TracePrefixed) {
+        self.nonverifiable_put_raw(NATIVE_ASSET_KEY.to_vec(), asset.to_string().into_bytes());
+    }
+
+    #[instrument(skip_all)]
+    fn put_ibc_asset(&mut self, asset: &TracePrefixed) -> Result<()> {
```

Comment:
The nomenclature here is bad, but the idea here is that this inserts a trace-prefixed asset into the state, such that sequencer can go `ibc/<hash>` -> TracePrefixed`.

I had opted to not rename anything in here because it really was just moving code around.